presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -532,6 +521,16 @@ public PlanOptimizers(
                 costCalculator,
                 new TranslateExpressions(metadata, sqlParser).rules()));
 
+        // TODO: Do not move other PlanNode to SPI until this rule is moved to the end of logical planning (i.e., where AddExchanges lives)
+        // TODO: The connector optimizer should not change plan shape (computations) at this point. (e.g., connector may pushdown filters to table scan but cannot add a new node)
```

Comment:
Is this accurate? The test has a rule that adds a filter node.

Reply:
We expected to land #12960 soon (after #12747) in a week or two. So after that, this TODO can be removed. Do we feel ok to leave as is given the PR is landing shortly. I will update the comment to indicate we need to move the connector rules above `HashGenerationOptimizer` as soon as we have merged #12960.